yb-voyager/cmd/importData.go

```diff
@@ -684,6 +699,60 @@ func importData(importFileTasks []*ImportFileTask) {
 
 }
 
+func importTasksViaTaskPicker(pendingTasks []*ImportFileTask, state *ImportDataState, progressReporter *ImportDataProgressReporter, poolSize int) error {
+	// The code can produce `poolSize` number of batches at a time. But, it can consume only
+	// `parallelism` number of batches at a time.
+	batchImportPool = pool.New().WithMaxGoroutines(poolSize)
+	log.Infof("created batch import pool of size: %d", poolSize)
+
+	taskPicker, err := NewSequentialTaskPicker(pendingTasks, state)
+	if err != nil {
+		return fmt.Errorf("create task picker: %w", err)
+	}
+	taskImporters := map[int]*FileTaskImporter{}
+
+	for taskPicker.HasMoreTasks() {
+		task, err := taskPicker.NextTask()
+		if err != nil {
+			return fmt.Errorf("get next task: %w", err)
+		}
+		var taskImporter *FileTaskImporter
+		var ok bool
+		taskImporter, ok = taskImporters[task.ID]
+		if !ok {
+			taskImporter, err = NewFileTaskImporter(task, state, batchImportPool, progressReporter)
+			if err != nil {
+				return fmt.Errorf("create file task importer: %s", err)
+			}
+			log.Infof("created file task importer for table: %s, task: %v", task.TableNameTup.ForOutput(), task)
+			taskImporters[task.ID] = taskImporter
+		}
+
+		if taskImporter.AllBatchesSubmitted() {
+			// All batches for this task have been submitted.
+			// task could have been completed (all batches imported) OR still in progress
+			// in case task is done, we should inform task picker so that we stop picking that task.
+			taskDone, err := taskImporter.AllBatchesImported()
+			if err != nil {
+				return fmt.Errorf("check if all batches are imported: task: %v err :%w", task, err)
+			}
+			if taskDone {
+				taskImporter.PostProcess()
+				err = taskPicker.MarkTaskAsDone(task)
+				if err != nil {
+					return fmt.Errorf("mark task as done: task: %v, err: %w", task, err)
+				}
+			}
+			continue
+		}
+		err = taskImporter.SubmitNextBatch()
+		if err != nil {
+			return fmt.Errorf("submit next batch: task:%v err: %s", task, err)
+		}
```

Comment:
What happens if all the batches are submitted but still in progress?
In that case there won't be any batch for `SubmitNextBatch()` but we are also not waiting here for batches to be imported and then mark the task as done by the picker.

Reply:
should we be waiting for the task import to be complete here?

Reply:
Yeah, good point, I also thought about that while writing this, but was worried if we might unnecessarily wait for too long. But I think I can add a sleep of 1 second perhaps before continuing (and checking again essentially). Makes sense?

Reply:
I might have misunderstood your question. 
we do wait for batches to be imported and only then mark the task as done by the picker. (see line 735, 739)

Reply:
line 735: taskImporter.AllBatchesImported()
This will just check, not wait for the batches to be imported ??

Reply:
Correct, but if all batches are not imported, it will not call `taskPicker.MarkTaskDone()`, we will simply `continue` (line 746)

Reply:
Ok, so `taskPicker.NextTask()` will keep returning the same task until we mark it as done, right?

Just thinking out loud, this loop will continuously keep checking, with any sleep, the state of task importer.
Should we add some sleep here to give the importer time to complete the in progress batch.

Reply:
Also, important thing to observe in this case(sleep vs no sleep) and overall also is - cpu consumption of voyager client machine.

Previously we used to have very low (less 50 or 10% mostly) cpu usage on client, iirc.
Check if there is any regression in that after refactoring.

Reply:
> Should we add some sleep here to give the importer time to complete the in progress batch.
Correct, that's what I thought you were referring to earlier. 

The problem is the sleep only makes sense for the sequentialTaskPicker, for the colocatedAwareRandomPicker, we should not sleep, we should just move on to the next task. Let me figure out a way to introduce that sleep only for SequentialTaskPicker.

Reply:
done.

Reply:
>Previously we used to have very low (less 50 or 10% mostly) cpu usage on client, iirc.
Check if there is any regression in that after refactoring.

Did a quick test to check this. Both main branch and this branch have similar characteristics:
It is mostly under 10, but it's also a bit spiky, i.e. when we move from one table to the next table, it briefly spikes up to 50-80% before coming down again. I also added a sleep, so don't think we should be much different in CPU consumption now.

Reply:
great, thanks.